### ISSUE TITLE
Bad docs SEO

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#52256]
Query `masked_fill_`:
![image](https://user-images.githubusercontent.com/1041752/107863539-822a0700-6e55-11eb-9308-65331a6fca30.png)
![image](https://user-images.githubusercontent.com/1041752/107863542-8b1ad880-6e55-11eb-9746-f46ffdd6b47c.png)
![image](https://user-images.githubusercontent.com/1041752/107863547-9241e680-6e55-11eb-97c6-e307f772b233.png)

Lot's of "spam" copy-paste sites like programmersthought, gitmemory, codesuche, kite, even github issues and the PyTorch one even does not highlight or scroll to https://pytorch.org/docs/stable/tensors.html?highlight=masked_fill_#torch.Tensor.masked_fill_

Maybe it would be best to have single-page per function just for SEO needs (even if the giant summary one continues to exist)

**Comment by imaginary-person** [comment]
I'm not a PyTorch team member, but I think another reason the SEO is currently lacking is because the docs automatically generated via Sphinx don't have HTML `meta` tags, such as `description` & `keywords`.  Sphinx [now supports](https://github.com/sphinx-doc/sphinx/pull/2516) adding these tags in documentation docs for better SEO.

**Comment by vadimkantorov** [comment]
This is a valid suggestion for single-functoin pages.

But for giant function collection page such as `tensors.html`, that unfortunately is unlikely to help :(

**Comment by imaginary-person** [comment]
I completely agree with you that the issue you posted can be solved with your proposed solution. My observation is a digression from this issue.

**Comment by vadimkantorov** [comment]
cc @zou3519

**Comment by zou3519** [comment]
Hmm, we already moved to single-page-per-function for torch.* functions. We should probably apply the same for Tensor.* functions

**Comment by vadimkantorov** [comment]
I think there should be for all functions detailed single-pages (peppered with enough of micro-formats to hide method signatures from google search snippets), and summary fast-loading table pages with very short function description for easier discoverability of all available functions.

Google is really spammed these days  with websites crawling from github issues and such, so good markup of docs is a must, otherwise other spam sites are winning the top-hits

**Comment by imaginary-person** [comment]
`Google is really spammed these days with websites crawling from github issues and such, so good markup of docs is a must, otherwise other spam sites are winning the top-hits`

I think this is where meta tags might also help.

**Comment by mruberry** [comment]
cc @brianjo

**Comment by zou3519** [comment]
To pull on this thread a little more, it sounds like there are two suggestions. The first is that we should split up all (or most) of our doc pages (vadimkantorov's suggestion). The second is, after splitting up the pages, populate the html with meta keyword and description tags to improve search indexing.

Split up PyTorch docs (below are the pages we should focus on first, but there are a lot more than these)
- [ ] Split torch.nn.functional docs
- [ ] Split torch.Tensor docs
- [ ] Split torch.autograd docs
- [ ] Split torch.optim docs
- [ ] Split torch.cuda docs
- [ ] Spilt torch.linalg, torch.fft docs
- [ ] Long-term: figure out what it would take to autosummary new docs pages by default. Maybe this is just adding a section to contributing.md.

Adding meta tags
- [ ] Add description and keywords meta tags to Sphinx HTML

**Comment by vadimkantorov** [comment]
Maybe even check HTML of the spam sites to imitate their meta-tag usage :)

**Comment by vadimkantorov** [comment]
Maybe the package pages could remain as concise tables listing all the functions (and invisible for search)

**Comment by vadimkantorov** [comment]
+ there's a giant tensor class, maybe worth hide its function lists, and make searchable only individual tensor.method pages

**Comment by zou3519** [comment]
cc @mattip

**Comment by ydcjeff** [comment]
Speaking of search, I am aware that pytorch is using algolia docsearch in https://pytorch.org front page and hub page.
So, can we bring that docsearch to https://pytorch.org/docs to make search a little more interactive ?

**Comment by mattip** [comment]
Please open a separate issue for algolia search.

**Comment by mattip** [comment]
I think we can close this. When searching google.com for `masked_fill_`, the new page shows up at number 4. Once 1.9 comes out and the canonical link to `masked_fill_` will be the smaller page, google should rank it higher.

There are still two bits that were discussed here:
- Long-term: figure out what it would take to autosummary new docs pages by default. Maybe this is just adding a section to contributing.md.
- Add description and keywords meta tags to Sphinx HTML

I think these should be separate issues. I once opened PR gh-38629 to refactor the community documentation including `contributing.md` but it did not gain traction. It would be good to add meta tags to the documentation, but that is a much bigger project than just reformatting the documentation to improve SEO scores.

**Comment by vadimkantorov** [comment]
Could you also check the html of the spam pages? Do they have any more magic meta tags?

**Comment by vadimkantorov** [comment]
Is snippet now better? Have you found way to block some parts of page from appearing in the snippets?

**Comment by vadimkantorov** [comment]
Are function list summary pages somehow made to not appear in the search?

**Comment by mattip** [comment]
@vadimkantorov you can check the google search results https://www.google.com/search?q=masked_fill_ . The top results (for me just now) are:
1. kite.com https://www.kite.com/python/docs/torch.Tensor.masked_fill_
2. a pytorch issue https://github.com/pytorch/pytorch/issues/39639
3. the page from the stable documentation, which has the "canonical" tag https://pytorch.org/docs/stable/tensors.html
4. the new page from master that will become the "canonical" one in two months when 1.9 is released https://pytorch.org/docs/master/generated/torch.Tensor.masked_fill_.html
5. A stackoverflow answer

I hope that once 1.9 is released, that google will re-index the canonical page and make it the top search hit. I am not very well versed in SEO terminology and do not know what you mean by "magic meta tags". Could you be more specific or point out values we could add to the page headers?

If you click on the link in (4) you can see the result from the newly rendered documentation, which is a specific page for that function. The [Tensor class documentation](https://pytorch.org/docs/master/tensors.html#tensor-class-reference) is now a table of method-summary, where the method is a link to that single page.

**Comment by vadimkantorov** [comment]
The new page's snippet is still bad: `masked_fill. Tensor. masked_fill (mask, value) → Tensor. Out-of-place version of ...`. Could we hide the signature from the snippets? Ideally, it should use some description meta tag.

![image](https://user-images.githubusercontent.com/1041752/115714462-6d974d00-a377-11eb-97db-1197670e2610.png)

**Comment by mattip** [comment]
Hmm. Given this docstring 
https://github.com/pytorch/pytorch/blob/56b43f4fec1f76953f15a627694d4bba34588969/torch/_tensor_docs.py#L4522-L4527

I guess we could generate a description tag as below by ignoring the first line, but I am not sure that is more helpful than the signature.
```
.. meta::
   :description: Out-of-place version of torch.Tensor.masked_fill_()
```

Is figuring out how to create better description meta tags  in scope for this issue (which is about getting the pytorch documentation into the top search results)?

**Comment by vadimkantorov** [comment]
Well, the concern about bad snippets/descriptions was in my first OP also. Would description tags fix it? Or can some parts of the page be annotated somehow to disallow appearance in snippets?

**Comment by vadimkantorov** [comment]
> but I am not sure that is more helpful than the signature.

Textual description should be more useful. (often cut or incomplete) signatures don't look nice in Google snippets and usually don't add uch information

**Comment by vadimkantorov** [comment]
The side panel and menus should be excluded from searchable text. I hope google provides some microformats to achieve that (https://developers.google.com/custom-search/docs/structured_data)
<img width="550" alt="image" src="https://user-images.githubusercontent.com/1041752/234993847-bbbd94c8-6423-4122-8d7c-b61fe853ae18.png">

**Comment by mattip** [comment]
The text "Built with Sphinx using a theme provided by Read the Docs. torch.Tensor.masked_fill. Docs." is still the top result for `masked_fill` on google.
